dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,634 @@
+      const auto& filter_shape = filter_shapes[sample_idx];
+      const auto& anchor_view = anchors[sample_idx];
+      assert(anchor_view.shape.num_elements() == filter_ndim);
+      int required_workspace;
+      bool has_degenerated_extents;
+      auto shape_desc =
+          SetupSampleShapeDesc(required_workspace, has_degenerated_extents, sample_idx, in_shape,
+                               filter_shape, anchor_view, shared_mem_limit);
+      any_has_degenerated_extents |= has_degenerated_extents;
+      max_total_workspace = std::max(max_total_workspace, required_workspace);
+      samples_desc_.push_back({out.tensor_data(sample_idx), in.tensor_data(sample_idx),
+                               filters.tensor_data(sample_idx), shape_desc});
+    }
+    int max_width = 0, max_height = 0;
+    ComputeExtentsMax(max_height, max_width, out.shape);
+    SampleDescT* descs_dev;
+    std::tie(descs_dev) = ctx.scratchpad->ToContiguousGPU(ctx.gpu.stream, samples_desc_);
+    int num_blocks_h = div_ceil(max_height, lanes);
+    int num_blocks_w = div_ceil(max_width, block_width);
+    num_blocks_h = std::min(num_blocks_h, max_grid_height);
+    num_blocks_w = std::min(num_blocks_w, max_grid_width);
+    dim3 grid(num_blocks_w, num_blocks_h, num_samples);
+    dim3 block(block_width, 1, 1);
+    RunKernelWithBorderMode(
+        ctx, border_type, fill_values, any_has_degenerated_extents,
+        [&](auto&& loader, auto&& roi_factory) {
+          filter::filter2d<<<grid, block, max_total_workspace, ctx.gpu.stream>>>(descs_dev, loader,
+                                                                                 roi_factory);
+          CUDA_CALL(cudaGetLastError());
+        });
+  }
+
+ protected:
+  template <typename KernelLauncher>
+  void RunKernelWithBorderMode(KernelContext& ctx, boundary::BoundaryType border_type,
+                               const TensorListView<StorageGPU, const In, 0>& fill_values,
+                               bool has_degenerated_extents, KernelLauncher&& launch_kernel) {
+    using namespace boundary;  // NOLINT(build/namespaces)
+    switch (border_type) {
+      case BoundaryType::REFLECT_101:
+        RunKernelBorder101(ctx, has_degenerated_extents, std::move(launch_kernel));
+        break;
+      case BoundaryType::REFLECT_1001:
+        RunKernelBorderRemap<filter::ROIFull, filter::Reflect1001>(ctx, std::move(launch_kernel));
+        break;
+      case BoundaryType::CLAMP:
+        RunKernelBorderRemap<filter::ROIFull, filter::Clamp>(ctx, std::move(launch_kernel));
+        break;
+      case BoundaryType::WRAP:
+        RunKernelBorderRemap<filter::ROIFull, filter::Wrap>(ctx, std::move(launch_kernel));
+        break;
+      case BoundaryType::CONSTANT:
+        RunKernelBorderConstant(ctx, fill_values, std::move(launch_kernel));
+        break;
+      case BoundaryType::ISOLATED:
+        RunKernelBorderRemap<filter::ROIOnlyValid>(ctx, std::move(launch_kernel));
+        break;
```

Comment:
ROI handling (including this one) can be achieved by specifying the output as smaller than input and appropriate anchors. To just use the pixels that have full halos present, you'd go with `output_size[d] = input_size[d] - kernel_size[d] + 1` and anchored at (0, 0). Note that this mechanism allows you to use totally arbitrary ROI.

Reply:
Done.